### PR TITLE
Add a generic OAM capabilities registration function with retries

### PR DESCRIPTION
**Pull request by utkarsh-pro**
Signed-off-by: Utkarsh Srivastava <srivastavautkarsh8097@gmail.com>

**Description**
This PR adds a generic OAM capabilities function with retries. This is being done so that capabilities registration process becomes more robust. 
Function internally uses exponential backoff and retries for 10 mins, after which it fails finally.

**Notes for Reviewers**
A new release of meshery-adapter-library after merging this PR would help move istio adapter OAM capabilities ahead.

**[Signed commits](../blob/master/CONTRIBUTING.md#signing-off-on-commits-developer-certificate-of-origin)**
- [x] Yes, I signed my commits.

### CODE REVIEW

File: adapter/oam.go [hunk 1]
```diff
@@ -1,8 +1,133 @@
...
+
+	backoff "github.com/cenkalti/backoff/v4"
+)
 
 // ProcessOAM processes OAM components. This is adapter specific and needs to be implemented by each adapter.
 func (h *Adapter) ProcessOAM(context.Context, OAMRequest) (string, error) {
 	return "", nil
 }
+
+// OAMRegistrant provides utility functions for registering
```
Comment (leecalcote) [inline adapter/oam.go:20]: The bifurcation of treatment of oam-operations and current operations is concerning: one is registered, while the other is discovered.

File: adapter/oam.go [hunk 2]
```diff
@@ -1,8 +1,133 @@
...
+	OAMHTTPRegistry string
+}
+
+// OAMRegistrantDefintionPath - Structure for configuring registrant paths
+type OAMRegistrantDefintionPath struct {
+	// OAMDefinitionPath holds the path for OAM Definition file
+	OAMDefintionPath string
+	// OAMRefSchemaPath holds the path for the OAM Ref Schema file
+	OAMRefSchemaPath string
+	// Host is the address of the gRPC host capabale of processing the request
```
Comment (leecalcote) [inline adapter/oam.go:40]:
```suggestion
	// Host is the address of the gRPC host capable of processing the request
```

File: adapter/oam.go [hunk 3]
```diff
@@ -1,8 +1,133 @@
...
+	Paths []OAMRegistrantDefintionPath
+
+	// OAMHTTPRegistry is the address of an OAM registry
+	OAMHTTPRegistry string
+}
+
+// OAMRegistrantDefintionPath - Structure for configuring registrant paths
+type OAMRegistrantDefintionPath struct {
+	// OAMDefinitionPath holds the path for OAM Definition file
+	OAMDefintionPath string
```
Comment (leecalcote) [inline adapter/oam.go:37]: "Definition"

File: adapter/oam.go [hunk 4]
```diff
@@ -1,8 +1,133 @@
...
+	// Host is the address of the gRPC host capabale of processing the request
+	Host string
+}
+
+// OAMRegistrantData struct defines the body of the POST request that is sent to the OAM
+// registry (Meshery)
+//
+// The body contains the
+// 1. OAM definition, which is in accordance with the OAM spec
+// 2. OAMRefSchema, which is json schema draft-4, draft-7 or draft-8 for the corresponding OAM object
```
Comment (leecalcote) [inline adapter/oam.go:49]: Talk nerdy to me, baby.

File: adapter/oam.go [hunk 5]
```diff
@@ -1,8 +1,133 @@
...
+		backoffOpt.MaxElapsedTime = 10 * time.Minute
+		if err := backoff.Retry(func() error {
+			contentByt, err := json.Marshal(ord)
+			if err != nil {
+				return backoff.Permanent(err)
+			}
+			content := bytes.NewReader(contentByt)
+
+			// host here is given by the application itself and is trustworthy hence,
+			// #nosec
```
Comment (leecalcote) [inline adapter/oam.go:110]: No admission control on "who" can register a capability, correct?

Comment (utkarsh-pro) [reply adapter/oam.go:110]: Nope, not yet. Maybe something to be built into meshery server?

File: adapter/oam.go (hunk 4 above)
Comment (utkarsh-pro) [reply adapter/oam.go:49]: Gotta be precise :stuck_out_tongue:

**Review by leecalcote — APPROVED** [review]
Good to go. Nice work!

**Comment by utkarsh-pro** [comment]
Can we get this PR moving? Actually it's blocking another PR on istio adapter. // @leecalcote @kumarabd @mgfeller @Aisuko

**Pull request merged**